TOOLS/lua/autocrop.lua

```diff
@@ -1,114 +1,344 @@
-crop_label = string.format("%s-crop", script_name)
 
--- number of seconds to gather cropdetect data
-detect_seconds = tonumber(mp.get_opt(string.format("%s.detect_seconds", script_name)))
-if not detect_seconds then
+local options = {
+    auto = true,
+    auto_delay = 4,
+    detect_limit = 24,
+    detect_round = 2,
+    detect_min_ratio = 0.5,
     detect_seconds = 1
-end
+}
+read_options(options)
 
-function del_filter_if_present(label)
-    -- necessary because mp.command('vf del @label:filter') raises an
-    -- error if the filter doesn't exist
-    local vfs = mp.get_property_native("vf")
-    for i,vf in pairs(vfs) do
-        if vf["label"] == label then
-            table.remove(vfs, i)
-            mp.set_property_native("vf", vfs)
+
+local label_prefix = mp.get_script_name()
+local labels = {
+    crop = string.format("%s-crop", label_prefix),
+    cropdetect = string.format("%s-cropdetect", label_prefix)
+}
+
+timers = {
+    auto_delay = nil,
+    detect_crop = nil
+}
+
+
+function is_filter_present(label)
+    local filters = mp.get_property_native("vf")
+    for index, filter in pairs(filters) do
+        if filter["label"] == label then
             return true
         end
     end
     return false
 end
 
-function autocrop_start()
-    -- exit if cropdetection is already in progress
-    if timer then
-        mp.msg.warn("already cropdetecting!")
-        return
+
+function is_time_enough(seconds)
+
+    -- Plus 1 second for deviation.
+    local time_needed = seconds + 1
+
+
+    -- For still images
+    if not is_playing_video() then
+        return true
     end
 
-    -- if there's a crop filter, remove it and exit
-    if del_filter_if_present(crop_label) then
+    return time_needed < mp.get_property_native("playtime-remaining")
+end
+
+
+function is_playing_video()
+
+    local vid = mp.get_property_native("vid")
+    local is_not_album = false == mp.get_property_native(
+        string.format("track-list/%s/albumart", vid)
+    )
+
+    return vid and is_not_album
+end
+
+
+function remove_filter(label)
+    if is_filter_present(label) then
+        mp.command(string.format('vf remove @%s', label))
+        return true
+    end
+    return false
+end
+
+
+function cleanup()
+
+    -- Remove all existing filters.
+    for key, value in pairs(labels) do
+        remove_filter(value)
+    end
+
+    -- Kill all timers.
+    for index, timer in pairs(timers) do
+        if timer then
+            timer:kill()
+            timer = nil
+        end
+    end
+end
+
+
+function detect_crop()
+
+    -- Verify if there is enough time for detect crop.
+    local time_needed = options.detect_seconds
+
+    if not is_time_enough(time_needed) then
+        mp.msg.warn("Not enough time for detect crop.")
         return
     end
 
-    -- insert the cropdetect filter
-    ret=mp.command(
+
+    -- Insert the cropdetect filter.
+    local limit = options.detect_limit/255
```

Comment:
The division by 255 should be in the default declaration, not here so that behaviour is the same as the filter itself. Also, there should be spaces around "/".

Reply:
I don’t think it ’s good for users to be able to enter an integer.

https://ffmpeg.org/ffmpeg-filters.html#cropdetect
0-255 only work right with 8bit.
a value between 0.0 and 1.0 will be scaled depending on the bitdepth of the pixel format.

This problem was fixed in this commit.
https://github.com/mpv-player/mpv/commit/1171f92beb4fd89e1ba4e76c9b0accacca36a32d#diff-4853f779c4417064fae59e54a43db8ec

Here is a wrong detect result with 10bit video.
```
string.format(
        'vf pre @%s:cropdetect=limit=%s:round=2:reset=0',
        cropdetect_label, 24
)
```
![image](https://user-images.githubusercontent.com/50797982/82863989-1960ce80-9f56-11ea-80d2-782bea2b6fb0.png)

```
string.format(
        'vf pre @%s:cropdetect=limit=%f:round=2:reset=0',
        cropdetect_label, 24/255
)
```
![image](https://user-images.githubusercontent.com/50797982/82864066-35647000-9f56-11ea-8d83-c3163617d645.png)

Reply:
But you are forcing users to enter an integer which is different that the documented filter behavior. I understand that you need to use a value between 0 and 1 for consistent results, so make the default "24 / 255", but if the user wants to use "0.1" or alternately rely on the documented scaling with bit depth, they should be able to do that.

Reply:
0-1 is a subset of 0-255 and cannot use fractional number in mpv.conf. 
add `detect_limit_den` to options?